pulsar-broker/src/main/java/org/apache/pulsar/client/impl/RawReaderImpl.java

```diff
@@ -203,12 +232,17 @@ public CompletableFuture<Void> seekAsync(MessageId messageId) {
 
         @Override
         public CompletableFuture<Void> closeAsync() {
+            CompletableFuture<Void> closeFuture = super.closeAsync();
             reset();
-            return super.closeAsync();
+            return closeFuture;
```

Comment:
`closeAsync` and `messageReceived` don't in the same thread run and have no locks. so the message also can be added in incomingRawMessages after it has been closed

step : 
1. thread-1 `messageReceived` check the state is ready.
https://github.com/apache/pulsar/pull/18928/files#diff-2d41ea4027219e13a0a942edc54352bb0a997060112705796d1d811213d585f3R245
2.  thread-2 `closeAsync()` the RawReaderImpl state change to close and reset the incomingRawMessages
https://github.com/apache/pulsar/pull/18928/files#diff-2d41ea4027219e13a0a942edc54352bb0a997060112705796d1d811213d585f3R245
3. thread-1 `messageReceived`  continue to run the add operation
https://github.com/apache/pulsar/pull/18928/files#diff-2d41ea4027219e13a0a942edc54352bb0a997060112705796d1d811213d585f3R252

Reply:
Yes, that's true. Pulsar contains a lot of such race conditions. I wonder if it's worth fixing that one, since it would add complexity to the code. There are multiple of such race conditions in ConsumerBase's closeAsync method. The fix would have to be done there. The problem is more severe in ConsumerBase since it can impact the memory limit controller. I'll create a separate issue to track the issue.

Reply:
I filed #18938

Reply:
I also added #18939